sw/device/lib/dif/dif_spi_device.c

```diff
@@ -0,0 +1,695 @@
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/dif/dif_spi_device.h"
+
+#include "spi_device_regs.h"  // Generated.
+#include "sw/device/lib/base/bitfield.h"
+#include "sw/device/lib/base/memory.h"
+
+const uint16_t kDifSpiDeviceSramLen = SPI_DEVICE_BUFFER_SIZE_BYTES;
+
+/**
+ * Computes the required value of the control register from a given
+ * configuration.
+ */
+static dif_spi_device_result_t build_control_word(
+    const dif_spi_device_config_t *config, uint32_t *device_config) {
+  *device_config = 0;
+  switch (config->clock_polarity) {
+    case kDifSpiDeviceEdgePositive:
+      *device_config |= 0x0 << SPI_DEVICE_CFG_CPOL;
+      break;
+    case kDifSpiDeviceEdgeNegative:
+      *device_config |= 0x1 << SPI_DEVICE_CFG_CPOL;
+      break;
+    default:
+      return kDifSpiDeviceResultBadArg;
+  }
+  switch (config->data_phase) {
+    case kDifSpiDeviceEdgePositive:
+      *device_config |= 0x1 << SPI_DEVICE_CFG_CPHA;
+      break;
+    case kDifSpiDeviceEdgeNegative:
+      *device_config |= 0x0 << SPI_DEVICE_CFG_CPHA;
+      break;
+    default:
+      return kDifSpiDeviceResultBadArg;
+  }
+  switch (config->tx_order) {
+    case kDifSpiDeviceBitOrderMsbToLsb:
+      *device_config |= 0x0 << SPI_DEVICE_CFG_TX_ORDER;
+      break;
+    case kDifSpiDeviceBitOrderLsbToMsb:
+      *device_config |= 0x1 << SPI_DEVICE_CFG_TX_ORDER;
+      break;
+    default:
+      return kDifSpiDeviceResultBadArg;
+  }
+  switch (config->rx_order) {
+    case kDifSpiDeviceBitOrderMsbToLsb:
+      *device_config |= 0x0 << SPI_DEVICE_CFG_RX_ORDER;
+      break;
+    case kDifSpiDeviceBitOrderLsbToMsb:
+      *device_config |= 0x1 << SPI_DEVICE_CFG_RX_ORDER;
+      break;
+    default:
+      return kDifSpiDeviceResultBadArg;
+  }
+
+  *device_config = bitfield_set_field32(
+      *device_config, (bitfield_field32_t){
+                          .mask = SPI_DEVICE_CFG_TIMER_V_MASK,
+                          .index = SPI_DEVICE_CFG_TIMER_V_OFFSET,
+                          .value = config->rx_fifo_timeout,
+                      });
+
+  return kDifSpiDeviceResultOk;
+}
+
+dif_spi_device_result_t dif_spi_device_init(
+    mmio_region_t base_addr, const dif_spi_device_config_t *config,
+    dif_spi_device_t *spi) {
+  if (spi == NULL) {
+    return kDifSpiDeviceResultBadArg;
+  }
+  spi->base_addr = base_addr;
+
+  // NOTE: we do not write to any registers until performing all
+  // function argument checks, to avoid a halfway-configured SPI.
+
+  uint32_t device_config;
+  dif_spi_device_result_t control_error =
+      build_control_word(config, &device_config);
+  if (control_error != kDifSpiDeviceResultOk) {
+    return control_error;
+  }
+
+  uint16_t rx_fifo_start = 0x0;
+  uint16_t rx_fifo_end = config->rx_fifo_len - 1;
+  uint16_t tx_fifo_start = config->rx_fifo_len;
+  uint16_t tx_fifo_end = config->rx_fifo_len + config->tx_fifo_len - 1;
```

Comment:
Using `tx_fifo_start` instead of `config->rx_fifo_len` would be clearer.

Reply:
Done.